crates/subspace-networking/src/protocols/connected_peers/tests.rs

```diff
@@ -0,0 +1,272 @@
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    peer2.listen().await;
+    peer1.connect(&mut peer2).await;
+
+    loop {
+        select! {
+            _ = peer1.next_swarm_event().fuse() => {},
+            _ = peer2.next_swarm_event().fuse() => {},
+            _ = sleep(large_delay).fuse() => {
+                break;
+            }
+        }
+    }
+
+    assert!(!peer1.is_connected(peer2.local_peer_id()));
+    assert!(!peer2.is_connected(peer1.local_peer_id()));
+}
+
+#[tokio::test()]
+async fn test_connection_decision() {
+    let decision_timeout = Duration::from_millis(500);
+    let large_delay = Duration::from_millis(2000);
+
+    let mut peer1 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer2 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    peer2.listen().await;
+    peer1.connect(&mut peer2).await;
+
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), true);
+
+    loop {
+        select! {
+            _ = peer1.next_swarm_event().fuse() => {},
+            _ = peer2.next_swarm_event().fuse() => {},
+            _ = sleep(large_delay).fuse() => {
+                break;
+            }
+        }
+    }
+
+    assert!(peer1.is_connected(peer2.local_peer_id()));
+    assert!(peer2.is_connected(peer1.local_peer_id()));
+}
+
+#[tokio::test()]
+async fn test_connection_decision_symmetry() {
+    let decision_timeout = Duration::from_millis(500);
+    let large_delay = Duration::from_millis(2000);
+
+    let mut peer1 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    let mut peer2 = new_ephemeral(
+        decision_timeout,
+        Behaviour::<ConnectedPeersInstance>::new(Config {
+            decision_timeout,
+            ..Default::default()
+        }),
+    );
+
+    peer2.listen().await;
+    peer1.connect(&mut peer2).await;
+
+    peer1
+        .behaviour_mut()
+        .update_keep_alive_status(*peer2.local_peer_id(), true);
+    peer2
+        .behaviour_mut()
+        .update_keep_alive_status(*peer1.local_peer_id(), false);
+
+    loop {
+        select! {
+            _ = peer1.next_swarm_event().fuse() => {},
+            _ = peer2.next_swarm_event().fuse() => {},
+            _ = sleep(large_delay).fuse() => {
+                break;
+            }
+        }
+    }
+
+    assert!(!peer1.is_connected(peer2.local_peer_id()));
+    assert!(!peer2.is_connected(peer1.local_peer_id()));
```

Comment:
Why would they not remain connected if one of the peers still wants to?

Reply:
A peer can't force remote peers to maintain connections.

Reply:
Hm... then we have a problem I think. Connected peers doesn't really maintain peers in connected state! It only prevents local end of the connection from being closed, which as you just mentioned is not sufficient to keep connections established.

So what this seems to mean is that we will still have a lot of reconnections unless both peers are interested in each other. More specifically this means nodes will continuously disconnect from farmers over and over again because farmers are never interested in nodes.

Reply:
> Hm... then we have a problem I think. Connected peers doesn't really maintain peers in connected state! It only prevents local end of the connection from being closed, which as you just mentioned is not sufficient to keep connections established.

I'm not sure I understand the problem. Peers can drop connections at any time with or without protocols for many reasons unless two peers decide to maintain them for some time. 

> So what this seems to mean is that we will still have a lot of reconnections unless both peers are interested in each other. More specifically this means nodes will continuously disconnect from farmers over and over again because farmers are never interested in nodes.

We could have reconnections with the same peers if you meant that but it would be the case after failing connections with all the other known peers. 

Nodes will maintain connections with farmers because we have two instances of this protocol on the farmer and one of them maintains connections with nodes.